### PR TITLE
Rate metrics converter

### LINKED ISSUE #984: [heapster-scalability] Convert missing rate metrics from cumulatives
https://github.com/kubernetes/heapster/blob/master/sinks/gcm/rate_metrics.go

### PR DESCRIPTION
**Pull request by mwielgus**
Fixes #984 
Needs unit tests and manual verification.

cc: @piosz @timstclair

### CODE REVIEW

File: metrics/core/types.go [hunk 1]
```diff
@@ -119,6 +118,8 @@ func (this *LabeledMetric) GetValue() interface{} {
 }
 
 type MetricSet struct {
+	CreateTime     time.Time
+	ScrapeTime     time.Time
```
Comment (timstclair) [inline metrics/core/types.go]: What about metrics that could be scraped at different times? Volume / filesystem stats will be updated on a different schedule from other stats, as will UserDefinedMetrics.

File: metrics/processors/rate_calculator.go [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+				continue
+			}
+
+			for metricName, targetMetric := range this.rateMetricsMapping {
+				metricValNew, foundNew := newMs.MetricValues[metricName]
+				metricValOld, foundOld := oldMs.MetricValues[metricName]
+				if foundNew && foundOld {
+					if metricName == core.MetricCpuUsageRate.MetricDescriptor.Name {
+						// cpu/usage values are in nanoseconds; we want to have it in millicores (that's why constant 1000 is here).
+						newVal := 1000 * (metricValNew.IntValue - metricValOld.IntValue) /
```
Comment (timstclair) [inline metrics/processors/rate_calculator.go]: How about definining a scale somewhere (either with the RateMetricsMapping, or maybe as part of the `core.UnitsType` in the MetricsDescriptor), then you don't need to branch on the metric name here

File: metrics/processors/rate_calculator.go [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+							(newMs.ScrapeTime.UnixNano() - oldMs.ScrapeTime.UnixNano())
+
+						newMs.MetricValues[targetMetric.MetricDescriptor.Name] = core.MetricValue{
+							ValueType:  core.ValueInt64,
+							MetricType: core.MetricGauge,
+							IntValue:   newVal,
+						}
+
+					} else if targetMetric.MetricDescriptor.ValueType == core.ValueFloat {
+						newVal := 1000000000 * float32(metricValNew.IntValue-metricValOld.IntValue) /
```
Comment (timstclair) [inline metrics/processors/rate_calculator.go]: (nit) prefer `time.Second.Nanoseconds()` or `1e9`

File: metrics/core/types.go (hunk 1 above)
Comment (mwielgus) [inline metrics/core/types.go]: Scrape time is (at least now) used only for converting rate metrics to cumulative. Do you have any cumulative metric that is scraped at a very different time? This is an internal structure that can be changed in the future. Right now I'm trying to reduce a size of a single metric as we are adding couple more of them.

File: metrics/processors/rate_calculator.go (hunk 2 above)
Comment (mwielgus) [inline metrics/processors/rate_calculator.go]: Good idea - will take a look.

**Comment by mwielgus** [comment]
Unit test added

File: metrics/core/types.go (hunk 1 above)
Comment (timstclair) [inline metrics/core/types.go]: This should be ok for now, especially if it's internal-only.

**Comment by mwielgus** [comment]
Added couple other changes required for this to work:
* do not aggregate container cumulative metrics 
* print scrape time in logsink

**Comment by piosz** [comment]
LGTM

**Pull request merged**